src/lib/edf_schedule.c

```diff
@@ -0,0 +1,536 @@
+			future_task = task;
+			break;
+		}
+	}
+
+	/* tell caller about future task */
+	return future_task;
+}
+
+/* cancel and delete task from scheduler - won't stop it if already running */
+int schedule_edf_task_cancel(struct sof_task *task)
+{
+	struct edf_schedule_data *sch = *arch_edf_schedule_get();
+	uint32_t flags;
+	int ret = 0;
+
+	tracev_schedule("schedule_edf_task_cancel()");
+
+	spin_lock_irq(&sch->lock, flags);
+
+	/* check current task state, delete it if it is queued
+	 * if it is already running, nothing we can do about it atm
+	 */
+	if (task->state == SOF_TASK_STATE_QUEUED) {
+		/* delete task */
+		task->state = SOF_TASK_STATE_CANCEL;
+		list_item_del(&task->list);
+	}
+
+	spin_unlock_irq(&sch->lock, flags);
+
+	return ret;
+}
+
+static int _sch_edf_task(struct sof_task *task, uint64_t start,
+	uint64_t deadline)
+{
+	struct edf_schedule_data *sch = *arch_edf_schedule_get();
+	uint32_t flags;
+	uint64_t current;
+	uint64_t ticks_per_ms;
+	struct edf_task_pdata *edf_pdata;
+
+	edf_pdata = edf_sch_get_pdata(task);
+
+	tracev_schedule("_schedule_edf_task()");
+
+	spin_lock_irq(&sch->lock, flags);
+
+	/* is task already pending ? - not enough MIPS to complete ? */
+	if (task->state == SOF_TASK_STATE_PENDING) {
+		trace_schedule("_schedule_edf_task(), task already pending");
+		spin_unlock_irq(&sch->lock, flags);
+		return 0;
+	}
+
+	/* is task already queued ? - not enough MIPS to complete ? */
+	if (task->state == SOF_TASK_STATE_QUEUED) {
+		trace_schedule("_schedule_edf_task(), task already queued");
+		spin_unlock_irq(&sch->lock, flags);
+		return 0;
+	}
+
+	/* get the current time */
+	current = platform_timer_get(platform_timer);
+
+	ticks_per_ms = clock_ms_to_ticks(sch->clock, 1);
+
+	/* calculate start time - TODO: include MIPS */
+	if (start == 0)
+		task->start = current;
+	else
+		task->start = task->start + ticks_per_ms * start / 1000 -
+			PLATFORM_SCHEDULE_COST;
+
+	/* calculate deadline - TODO: include MIPS */
+	edf_pdata->deadline = task->start + ticks_per_ms * deadline / 1000;
+
+	/* add task to list */
+	list_item_append(&task->list, &sch->list);
+	task->state = SOF_TASK_STATE_QUEUED;
+	spin_unlock_irq(&sch->lock, flags);
+
+	return 1;
+}
+
+/*
+ * Add a new task to the scheduler to be run and define a scheduling
+ * deadline in time for the task to be ran. Do not invoke the scheduler
+ * immediately to run task, but wait intil schedule is next called.
+ *
+ * deadline is in microseconds relative to start.
+ */
+static void _schedule_edf_task_idle(struct sof_task *task, uint64_t deadline)
+{
+	_sch_edf_task(task, 0, deadline);
+}
+
+/*
+ * Add a new task to the scheduler to be run and define a scheduling
+ * window in time for the task to be ran. i.e. task will run between start and
+ * deadline times.
+ *
+ * start is in microseconds relative to last task start time.
+ * deadline is in microseconds relative to start.
+ */
+static void _schedule_edf_task(struct sof_task *task, uint64_t start,
```

Comment:
Minor style problems:
1) Having both `_schedule_edf_task` and `schedule_edf_task` may be misleading, if `schedule_edf_task` is just proxy/switch to other functions just add suffixes to them like for `_idle`, this one can be `_normal` or smth like that.
2) Why sometimes static functions are prefixed with `_` and sometimes not?

Reply:
1. I've rename _schedule_edf_task to schedule_edf_task_normal.
2. As it seems to me ```_``` prefix is not a "property" for all static functions. I've noticed that it is used for only for those, which are invoked by proxy/switch function.